src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker_IVsSolutionLoadEvents.cs

```diff
@@ -51,6 +95,217 @@ int IVsSolutionLoadEvents.OnAfterBackgroundSolutionLoadComplete()
+                cancellationToken).ConfigureAwait(true);
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
+            OutputToOutputWindow($"Getting project information - done (took {DateTimeOffset.UtcNow - start})");
+
+            OutputToOutputWindow($"Creating projects - start");
+            start = DateTimeOffset.UtcNow;
+            var targetPathsToProjectPaths = BuildTargetPathMap(projectInfos);
+            foreach (var projectFilename in projectInfos.Keys)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                CreateProjectFromArgumentsAndReferences(
+                    workspaceProjectContextFactory,
+                    projectFilename,
+                    projectInfos,
+                    targetPathsToProjectPaths);
+            }
+            OutputToOutputWindow($"Creating projects - done (took {DateTimeOffset.UtcNow - start})");
+
+            OutputToOutputWindow($"Pushing to workspace - start");
+            start = DateTimeOffset.UtcNow;
+            FinishLoad();
+            OutputToOutputWindow($"Pushing to workspace - done (took {DateTimeOffset.UtcNow - start})");
+        }
+
+        private static ImmutableDictionary<string, string> BuildTargetPathMap(IReadOnlyDictionary<string, DeferredProjectInformation> projectInfos)
+        {
+            var builder = ImmutableDictionary.CreateBuilder<string, string>();
+            foreach (var item in projectInfos)
+            {
+                var targetPath = item.Value.TargetPath;
+                if (targetPath != null)
+                {
+                    if (!builder.ContainsKey(targetPath))
+                    {
+                        builder[item.Value.TargetPath] = item.Key;
+                    }
+                    else
+                    {
+                        Debug.Fail($"Already have a target path of '{item.Value.TargetPath}', with value '{builder[item.Value.TargetPath]}'.");
+                    }
+                }
+            }
+            return builder.ToImmutable();
+        }
+
+        [Conditional("DEBUG")]
+        private void OutputToOutputWindow(string message)
+        {
+            _pane?.OutputString(message + Environment.NewLine);
+        }
+
+        private IWorkspaceProjectContext CreateProjectFromArgumentsAndReferences(
+            IWorkspaceProjectContextFactory workspaceProjectContextFactory,
+            string projectFilename,
+            IReadOnlyDictionary<string, DeferredProjectInformation> allProjectInfos,
+            IReadOnlyDictionary<string, string> targetPathsToProjectPaths)
+        {
+            var languageName = GetLanguageOfProject(projectFilename);
+            if (languageName == null)
+            {
+                return null;
+            }
+
+            DeferredProjectInformation projectInfo;
+            if (!allProjectInfos.TryGetValue(projectFilename, out projectInfo))
```

Comment:
If we're trying to satisfy a P2P ref from another project, but the target of the p2p ref doesn't actually exist.  Will comment.